tokio-util/src/task/join_map.rs

```diff
@@ -416,26 +396,26 @@ where
     }
 
     fn insert(&mut self, key: K, abort: AbortHandle) {
-        let hash = self.hash(&key);
+        let hash_builder = self.hashes_by_task.hasher();
+        let hash = hash_one(hash_builder, &key);
         let id = abort.id();
-        let map_key = Key { id, key };
 
         // Insert the new key into the map of tasks by keys.
-        let entry = self
-            .tasks_by_key
-            .raw_entry_mut()
-            .from_hash(hash, |k| k.key == map_key.key);
+        let entry =
+            self.tasks_by_key
+                .entry(hash, |(k, _)| *k == key, |(k, _)| hash_one(hash_builder, k));
         match entry {
-            RawEntryMut::Occupied(mut occ) => {
+            Entry::Occupied(occ) => {
                 // There was a previous task spawned with the same key! Cancel
                 // that task, and remove its ID from the map of hashes by task IDs.
-                let Key { id: prev_id, .. } = occ.insert_key(map_key);
-                occ.insert(abort).abort();
-                let _prev_hash = self.hashes_by_task.remove(&prev_id);
+                let (_, abort) = std::mem::replace(occ.into_mut(), (key, abort));
```

Comment:
This drops the key. If the destructor of the key panics, we don't call `abort` or adjust `hashes_by_task`.

Reply:
Good catch